### DISCUSSION TITLE
HashIndex values not dropped until the end of the program?

### DISCUSSION THREAD

**Discussion by Dreaming-Codes** [wvwwvwwv/scalable-concurrent-containers#136] [Q&A] [answered]
What's happening [here](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10Kc2NjID0gIjIuMC4xOCIKKi8KCnVzZSBzY2M6Okhhc2hJbmRleDsKCiNbZGVyaXZlKENsb25lLCBEZWJ1ZyldCnN0cnVjdCBUZXN0U3RydWN0KHB1YiBTdHJpbmcpOwoKaW1wbCBEcm9wIGZvciBUZXN0U3RydWN0IHsKICAgIGZuIGRyb3AoJm11dCBzZWxmKSB7CiAgICAgICAgcHJpbnRsbiEoInt9IGhhcyBiZWVuIGRyb3BwZWQuIiwgc2VsZi4wKQogICAgfQp9CgpmbiBtYWluKCkgewogICAgewogICAgICAgIGxldCB0ZXN0ID0gVGVzdFN0cnVjdCgibm9ybWFsIHRlc3QiLnRvX3N0cmluZygpKTsKICAgIH0gLy8gQXQgdGhpcyBwb2ludCBpbiB0aGUgZXhlY3V0aW9uLCAibm9ybWFsIHRlc3QgaGFzIGJlZW4gZHJvcHBlZC4iIHdpbGwgYmUgcHJpbnRlZC4KICAgIHByaW50bG4hKCJUaGUgJ25vcm1hbCB0ZXN0JyBzaG91bGQgaGF2ZSBiZWVuIGRyb3BwZWQgYnkgbm93LiIpOwogICAgLy8gTm93LCBsZXQncyB0cnkgaXQgd2l0aCBIYXNoSW5kZXguCiAgICB7CiAgICAgICAgbGV0IGhhc2hfaW5kZXg6IEhhc2hJbmRleDx1OCwgVGVzdFN0cnVjdD4gPSBIYXNoSW5kZXg6Om5ldygpOwogICAgICAgIAogICAgICAgIGhhc2hfaW5kZXguaW5zZXJ0KDAsIFRlc3RTdHJ1Y3QoImhhc2hfaW5kZXggdGVzdCIudG9fc3RyaW5nKCkpKTsKICAgICAgICBoYXNoX2luZGV4LnJlbW92ZSgmMCk7CiAgICB9CiAgICBwcmludGxuISgiVGhlICdoYXNoX2luZGV4IHRlc3QnIHNob3VsZCBoYXZlIGJlZW4gZHJvcHBlZCBiZWZvcmUgdGhpcyBtZXNzYWdlLCBub3QgYWZ0ZXIuIik7Cn0=)?

**Comment by Dreaming-Codes** [comment]
Temporary switched to HashMap since my app rely on the value getting dropped

**Reply by wvwwvwwv** [reply]
I guess HashIndex is not for you if you need to depend on an exact timing of ‘drop’ calls.

**Comment by wvwwvwwv** [comment, marked as answer]
Hi! Values in a HashIndex are not immediately dropped, and they may last longer than the HashIndex instance.
1. When an entry is removed, they are just marked deleted, because there might be lock-free readers of it. The instance is passed to the EBR garbage collector when the bucket is rearranged (frequent remove & insert triggers a rearrange), the HashIndex is resized, or the HashIndex is dropped.
2. The EBR garbage collector may not drop garbage instances until the end of the process.

**Reply by Dreaming-Codes** [reply]
> 1. When an entry is removed, they are just marked deleted, because there might be lock-free readers of it. The instance is passed to the EBR garbage collector when the bucket is rearranged (frequent remove & insert triggers a rearrange), the HashIndex is resized, or the HashIndex is dropped.

Is there a way to listen for deletion?

**Reply by wvwwvwwv** [reply]
No, and it's not just about knowing when the entry is dropped, but the fact that the entry can survive indefinitely. That's a cost for allowing lock-free read.

**Comment by wvwwvwwv** [comment]
If you have any other questions, don't hesitate to open a new discussion thread!
